src/apps/appstore/app.js

```diff
@@ -393,8 +471,296 @@ define(function(require) {
 					$('#appstore_container .app-filter.active').click();
 				});
 			});
-		}
+		},
+
+		showMarketplaceConnector: function(parent) {
+			const self = this;
+			const marketConfig = self.appstoreData.marketplace || {};
+			const enabled = marketConfig.enabled;
+			const is_linked = marketConfig.is_linked;
+
+			if (enabled && is_linked) {
+				self.renderMarketSettings(parent);
+			} else if (marketConfig.enabled) {
+				self.renderMarketLink(parent);
+			} else {
+				self.renderMarketWelcome(parent);
+			}
+		},
+
+		renderMarketLink: function(parent) {
+			const self = this;
+			const marketConfig = self.appstoreData.marketplace;
+			const template = $(self.getTemplate({
+				name: 'marketClusterLink',
+				data: {
+					config: marketConfig
+				}
+			}));
+			const linkForm = template.find('#cluster_link_form');
+
+			parent
+				.find('.right-container')
+				.fadeOut(function() {
+					$(this)
+						.empty()
+						.append(template)
+						.fadeIn();
+					monster.ui.validate(linkForm);
+					self.bindMarketLinkEvents(parent, template);
+				});
+		},
+
+		bindMarketLinkEvents: function(parent, template) {
+			const self = this;
+			const linkForm = template.find('#cluster_link_form');
+			const maybeSetApiUrl = function() {
+				if (!self.appstoreData.marketplace.api_url && monster.config.api.default) {
+					self.updateMarketConnector({
+						action: 'api_url',
+						api_url: monster.config.api.default
+					});
+				}
+			};
+
+			parent.find('.market-link #save').on('click', function(e) {
+				e.preventDefault();
+
+				const saveThis = $(this);
+
+				saveThis.prop({ disabled: true });
+
+				if (monster.ui.valid(linkForm)) {
+					const formData = monster.ui.getFormData('cluster_link_form');
+
+					maybeSetApiUrl();
+					self.updateMarketConnector(
+						{
+							action: 'link',
+							access_code: formData.access_code,
+							name: formData.cluster_name
+						},
+						function() {
+							self.showMarketplaceConnector(
+								parent
+							);
+						},
+						function() {
+							saveThis.prop({ 'disabled': false });
+						});
```

Comment:
One thing to note here, this code is executed asynchronously, meaning `self#updateMarketConnector` might start executing BEFORE `maybeSetApiUrl` is done executing.

If `self#updateMarketConnector` relies on `maybeSetApiUrl` being executed before it starts its own API request, these requests will have to be handled with a control flow.

Here, we basically want to run them in series (waiting for one to resolve before the next one starts):
```js
monster.series([
   maybeSetApiUrl,
   function(next) {
      self.updateMarketConnector(..., _.partial(next, null), _.partial(next, true))
   }
], function(err) {
   if (err) {
      saveThis.prop({ 'disabled': false });
      return;
   }
   self.showMarketplaceConnector(parent);
});
```
And `maybeSetApiUrl` will have to be modified to take in a callback:
```js
			const maybeSetApiUrl = function(next) {
				if (!self.appstoreData.marketplace.api_url && monster.config.api.default) {
					self.updateMarketConnector({
						action: 'api_url',
						api_url: monster.config.api.default
					}, _.partial(next, null));
				}
			};
```